### PR TITLE
Slack engine class fixes

### PR DESCRIPTION
**Pull request by garethgreenaway**
### What does this PR do?
Updating the current Slack engine in develop with various fixes from stable branch, including sending kwargs correctly, passing pillar values and editing previous messages.  Some additional cleanup and fixes, eg. swapping the engine to a class for ease of passing data around in various functions.

### What issues does this PR fix or reference?
N/A

### Previous Behavior
See above.

### New Behavior
See above.

### Tests written?
No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

### CODE REVIEW

**Comment by garethgreenaway** [comment]
Lint.  Update inbound.

File: salt/engines/slack.py [hunk 1]
```diff
@@ -129,600 +131,617 @@ def __virtual__():
...
+                'exclude_members': True
+            })
+        channels = {}
+        if 'message' in ret:
+            for item in ret['message']:
+                channels[item["id"]] = item["name"]
+        return channels
+
+    def get_config_groups(self, groups_conf, groups_pillar_name):
+        """
```
Comment (cachedout) [inline salt/engines/slack.py]: single quotes please

File: salt/engines/slack.py [hunk 2]
```diff
@@ -129,600 +131,617 @@ def __virtual__():
...
+                ret_groups[name]['default_target'].update(config.get('default_target', {}))
+                ret_groups[name]['targets'].update(config.get('targets', {}))
+            except IndexError:
+                log.warn("Couldn't use group {}. Check that targets is a dict and not a list".format(name))
+
+        log.debug("Got the groups: {}".format(ret_groups))
+        return ret_groups
+
+    def _groups_from_pillar(self, pillar_name):
+        """pillar_prefix is the pillar.get syntax for the pillar to be queried.
```
Comment (cachedout) [inline salt/engines/slack.py]: single quotes and newline

File: salt/engines/slack.py [hunk 3]
```diff
@@ -129,600 +131,617 @@ def __virtual__():
...
+        log.info("pillar groups type is {}".format(type(pillar_groups)))
+        if pillar_groups:
+            return pillar_groups
         else:
-            cmdlist.append(cmditem)
-    return cmdlist
+            return {}
+
+    def fire(self, tag, msg):
+        """
```
Comment (cachedout) [inline salt/engines/slack.py]: single quotes

File: salt/engines/slack.py [hunk 4]
```diff
@@ -129,600 +131,617 @@ def __virtual__():
...
+        else:
+            fire_master = None
 
+        if fire_master:
+            fire_master(msg, tag)
+        else:
+            __salt__['event.send'](tag, msg)
+
+    def can_user_run(self, user, command, groups):
+        """
```
Comment (cachedout) [inline salt/engines/slack.py]: single quotes

File: salt/engines/slack.py [hunk 5]
```diff
@@ -129,600 +131,617 @@ def __virtual__():
...
+            if (command not in val['commands']) and (command not in val.get('aliases', {}).keys()):
+                if '*' not in val['commands']:
+                    continue  # again, pass
+            log.info("Slack user {} permitted to run {}".format(user, command))
+            return (key, val,)  # matched this group, return the group
+        log.info("Slack user {} denied trying to run {}".format(user, command))
+        return ()
+
+    def commandline_to_list(self, cmdline_str, trigger_string):
+        """
```
Comment (cachedout) [inline salt/engines/slack.py]: single quotes

**Review by cachedout — CHANGES_REQUESTED** [review]
I marked some of the cases where comments need to be cleaned up but there are a few others as well.

**Comment by rallytime** [comment]
@cachedout re-review requested

File: salt/engines/slack.py [hunk 6]
```diff
@@ -129,600 +131,621 @@ def __virtual__():
...
+                ret_groups[name]['default_target'].update(config.get('default_target', {}))
+                ret_groups[name]['targets'].update(config.get('targets', {}))
+            except IndexError:
+                log.warn("Couldn't use group {}. Check that targets is a dict and not a list".format(name))
+
+        log.debug("Got the groups: {}".format(ret_groups))
+        return ret_groups
+
+    def _groups_from_pillar(self, pillar_name):
+        '''pillar_prefix is the pillar.get syntax for the pillar to be queried.
```
Comment (cachedout) [inline salt/engines/slack.py]: We need a newline here so this is easier to read.

File: salt/engines/slack.py [hunk 7]
```diff
@@ -129,600 +131,621 @@ def __virtual__():
...
+        in a jinja template.
+
+        returns a dictionary (unless the pillar is mis-formatted)
+        XXX: instead of using Caller, make the minion to use configurable so there could be some
+             restrictions placed on what pillars can be used.
+        '''
+        caller = salt.client.Caller()
+        pillar_groups = caller.cmd('pillar.get', pillar_name)
+        # pillar_groups = __salt__['pillar.get'](pillar_name, {})
+        log.info("Got pillar groups {} from pillar {}".format(pillar_groups, pillar_name))
```
Comment (cachedout) [inline salt/engines/slack.py]: Single-quotes here and on the next two lines please. :]

**Review by cachedout — CHANGES_REQUESTED** [review]
Some minor style issues. I didn't comment on the use of double-quotes in many other places in this change but please remember that our style-guide does indicate that they should be used over single-quotes. :]

File: salt/engines/slack.py (hunk 7 above)
Comment (garethgreenaway) [reply salt/engines/slack.py]: Both fixed.  These were all introduced in a previous commit and I believe I've caught them all in this last push.

**Comment by rallytime** [comment]
re-run lint

**Review by cachedout — APPROVED** [review]

**Pull request merged**